Support for Sony Bravia TV

antonioparraga opened the pull request:
**Description:**
The braviatv platform allows you to control a Sony Bravia TV.

To add a Sony Bravia TV to your installation, add the following to your configuration.yaml file:

``` yaml
# Example configuration.yaml entry
media_player:
  platform: braviatv
  host: 192.168.5.22
  name: Living Room TV
```

Configuration variables:
- host (Required): The IP of the Sony Bravia TV, eg. 192.168.5.22
- name (Optional): The name you would like to give to the Sony Bravia TV.

You will need to configure your TV to allow the Home Assistant for remote usage. To do that, ensure that your TV is turned on. Open the configuration popup on Home Assistant and enter whatever PIN (for example, 0000). After that, the TV will show you a PIN and the Home Assistant will allow you to re-enter a new PIN. Enter the PIN shown on your TV and your Home Assistant will be allowed for remote usage on your Sony Bravia TV.

Supported almost all Sony Bravia TV 2013 and newer at http://info.tvsideview.sony.net/en_ww/home_device.html#bravia

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
# Example configuration.yaml entry
media_player:
  platform: braviatv
  host: 192.168.5.22
  name: Living Room TV
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

antonioparraga committed:
Added Sony Bravia support to HA

antonioparraga committed:
Improvements to make it work on my poor raspberry 1

antonioparraga committed:
Just a typo

antonioparraga committed:
A few fixes in order to pass pylint

rmkraus commented:
This platform contains a lot of logic for communicating with the TV. It would be best to split this out into a separate library and add that library as a dependency.

Nice use of the configurator component, though. 👍  That one definitely goes under-appreciated.

aparraga commented:
Thank you regarding the configurator usage. I think that the configurator is a great piece for this kind of use cases. You should work on improving it to support more scenarios.
Regarding your comment, I don't see lot of logic within my component, just the necessary to communicate with bravia TV taking into account current HA capabilities. I have performed a big clean-up before merging this guy into your branch and I think we could save a dependency in that sense. 
We could think about moving out to a library if we have to support more devices using the same protocol.
The component work like a charm, I have it connected to my Amazon ECHO via HA and after going back from work I have just to tell it: "Alexa, tell home assistant to turn on the TV" :)

balloob commented on homeassistant/components/media_player/braviatv.py:
```diff
@@ -0,0 +1,521 @@
+# flake8: noqa
```
Nop

balloob commented:
Yeah.

We don't want any protocol parsing etc in this repo. That way other people that do not use HA will not be able to leverage your work and we will have to maintain protocol upgrades.

antonioparraga committed:
- Remove noqa: was due to the 80 characters max per line restriction
- Move communication logic to a separate library at https://github.com/aparraga/braviarc.git
- Added dependency and adapt the code according to that

aparraga commented:
Ok, the code has been updated according to your suggestions.
Thank you!

antonioparraga committed:
A few improvements

antonioparraga committed:
Just a typo in a comment

balloob commented:
It looks good. Please rebase on the latest `dev` branch to resolve your merge conflicts.

antonioparraga committed:
Rebase from HM/dev

aparraga commented:
Ok, done!

balloob commented:
You didn't rebase but merged the branch. Now it shows 23 files changed and I can no longer verify which changes this PR brings. Please fix it 👍

aparraga commented:
Ops, sorry! fixed!

antonioparraga committed:
Update requirements by executing the script/gen_requirements_all.py

balloob commented on homeassistant/components/frontend/www_static/home-assistant-polymer:
```diff
@@ -1 +1 @@
-Subproject commit 0e6dc25fcd09ad1150aab258f8d01491a8ee8db7
+Subproject commit 77f4dd1fed3d29c7ad8960c704a748af80748a59
```
Please remove this from your PR.

balloob commented on homeassistant/components/media_player/braviatv.py:
```diff
@@ -0,0 +1,473 @@
+        try:
+            with open(filename, 'w') as fdesc:
+                string_io = StringIO()
+                json.dump(new_config, string_io)
```
what? You can either pass in `fdesc` straight into `json.dump` or you call `json.dumps(…)` and get the string value which you pass to `fdesc.write(…)`. There is no need for `StringIO`.

balloob commented on homeassistant/components/media_player/braviatv.py:
```diff
@@ -0,0 +1,473 @@
+
+    def _refresh_volume(self):
+        resp = self. \
+            _braviarc.bravia_req_json(self._host,
```
Wouldn't it make sense to move such calls also to the Bravia package? That way you can call things like

```python
resp = self._braviarc.volume_info()
resp = self._braviarc.remote_controller_info()
# etc…
```

balloob commented on homeassistant/components/media_player/braviatv.py:
```diff
@@ -0,0 +1,473 @@
+
+    def media_next_track(self):
+        """Send next track command."""
+        self._braviarc.send_req_ircc(self._host,
```
These could also be moved to the bravia object.

```python
self._braviarc.send_next()
self._braviarc.send_next()
self._braviarc.send_play()
self._braviarc.send_prev()
self._braviarc.play_content(uri)
# Etc…
```

antonioparraga committed:
More isolation level for braviarc lib

antonioparraga committed:
Remove unnecessary StringIO usage

antonioparraga committed:
Revert submodule polymer commit

antonioparraga committed:
Small refactorization and clean up of unused functions

antonioparraga committed:
Executed script/gen_requirements_all.py

antonioparraga committed:
Added a missing condition to ensure that a map is not null

aparraga commented:
Ok, done! Please review it again to check if everything looks cool now!

antonioparraga committed:
Fix missing parameter detected by pylint

antonioparraga committed:
A few improvements, also added an empty line to avoid the lint error

antonioparraga committed:
A typo

balloob commented:
Yep, all cool! 🐬 ⛄ 🌴

dennuuz commented:
Hi Aparraga,

Many thanks for building this component! It's (almost) working great! I copied the code to my custom sensor folder for now. 

I have a Sony Bravia KDL myself, connected to my WIFI network. When my TV is turned on, everything works great, however, when it's turned off the below error message is logged. This is obviously caused by the braviarc submitting a request while the host is not available. 

I started to fiddle around a bit myself to see whether i can first ping the host before submitting the request but haven't been able to solve it so far. 

You help would be appreciated. 

`Jun 23 19:48:09 raspberrypi hass[5165]: ERROR:braviarc.braviarc:[W] Exception: HTTPConnectionPool(host='192.168.2.107', port=80): Max retries exceeded with url: /sony/accessControl (Caused by ConnectTimeoutError(<requests.packages.urllib3.connection.HTTPConnection object at 0x76426d70>, 'Connection to 192.168.2.107 timed out. (connect timeout=10)'))`

aparraga commented:
Thank you Dennuuz,

Problem with the ping is that Bravia responds to pings even if it's turned off.
So, component can't guarantee Bravia status based on ping. I think that checking the status of the TV by calling rest service is a good choice. Maybe component should catch this exception to avoid log the error.
Does it make sense?
